### PR TITLE
Detect use-after-scope bugs with AddressSanitizer

### PR DESCRIPTION
**Pull request by tmiasko**
Enable use-after-scope checks by default when using AddressSanitizer.
They allow to detect incorrect use of stack objects after their scope
have already ended. The detection is based on LLVM lifetime intrinsics.

To facilitate the use of this functionality, the lifetime intrinsics are
now emitted regardless of optimization level if enabled sanitizer makes
use of them.

### CODE REVIEW

File: src/test/ui/sanitizer-use-after-scope.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// needs-sanitizer-support
```
Comment (Centril) [inline src/test/ui/sanitizer-use-after-scope.rs]: Please move the test to a relevant sub-directory of `ui/` rather than having it directly in `ui/`.

Comment (tmiasko) [reply src/test/ui/sanitizer-use-after-scope.rs]: Sanitizer ui tests now have a separate subdirectory.

**Comment by nikic** [comment]
Does this need to be exposed as an option, similar to msan origin tracking? I know that it can be disabled via `ASAN_OPTIONS` at runtime, but wondering if there are any salient reasons to disable the instrumentation itself (like performance).

**Comment by tmiasko** [comment]
I think a new option is unnecessary. Use-after-scope has been enabled by
default in clang for some time now https://reviews.llvm.org/D31479. 

Of course, in the case anyone would like to have a flag, I can add one.

**Comment by nikic** [comment]
I think we're fine without the flag until someone has a use-case for it...

@bors r+

**Pull request merged**